### PR TITLE
Chat !

### PR DESCRIPTION
**Pull request by nushicoco**
includes : 
* button to chat with tech (the same for regular user/ anonymous and superuser)
* chat url where they can communicate 
* modal to ask anon user his nickname (missing validation and default enter)
* couple of socket tests and route tests
* ticket missing tests 
* introduced - socket.io, socket.io-client (both on client and server for tests), moment.js

still need to do:
* use env vars for react socket connect
* nicer empty no pending chats page
* modal validations
* some edge case bug (that will prolly require redux) = > if superuser is connected through direct link, writes some msgs and then refresh - old msgs will appear as not from him in the view (on the left and red)
* the implementation is solely for users to initiate a chat and only superusers can answer, we will prolly need to generalize it more to allow users working on the same project to talk with each other.

Halleluja

### CODE REVIEW

File: test/socket/chat.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    let clientSendMessage = `client:sendMessage`;
+    let serverConnected = `server:connected`;
+    let serverGotMessage = `server:gotMessage`;
+
+    var superUser = {name:"ella", id:1};
+    var chatClient1 = {name:"scrappy", id:2};
+
+    it('clientId as userId is returned correctly from server', function (done){
+
+        // this.timeout(1000);
```
Comment (dancar) [inline test/socket/chat.js]: ✂️

File: test/socket/chat.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    let clientConnected = `client:connected`;
+    let clientSendMessage = `client:sendMessage`;
+    let serverConnected = `server:connected`;
+    let serverGotMessage = `server:gotMessage`;
+
+    var superUser = {name:"ella", id:1};
+    var chatClient1 = {name:"scrappy", id:2};
+
+    it('clientId as userId is returned correctly from server', function (done){
```
Comment (dancar) [inline test/socket/chat.js]: Consider some more appropriate BDD 'storytelling': 
`it('should correctly return clientId as userId from server', ...)`

Same goes for the rest of the tests

File: test/socket/chat.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        // this.timeout(1000);
+
+        var client1 = io.connect(socketURL);
+
+        client1.emit(clientConnected, { chatId : 1 , client: chatClient1.name, clientId: chatClient1.id});
```
Comment (dancar) [inline test/socket/chat.js]: style - multiline objects are far more readable IMHO

File: test/socket/chat.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        client1.on (serverConnected, data => {
+
+            data.clientId.should.equal(chatClient1.id)
+            done();
+        });
+
+
+        // client1.on(`server:gotMessage`, data => {
```
Comment (dancar) [inline test/socket/chat.js]: ✂️

File: test/socket/chat.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        // })
+    });
+
+    it('clientId is generated by server when not sent by user', function (done){
+
+        // this.timeout(1000);
```
Comment (dancar) [inline test/socket/chat.js]: ✂️

File: client/src/components/chat/chat.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        })
+        .catch(function (err){
+            // todo what's here?
+        })
+    }
+
+    componentWillMount(){
+        if (this.state.user){
+            this.setState({client:this.state.user.firstName});
+            this.createChat();
```
Comment (nushicoco) [inline client/src/components/chat/chat.js]: I have a createChat on the onExited of the modal as well

File: client/src/components/chat/chat.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    handleChatChange (index) {
+        this.setState({activeChatIndex : index });
+    }
+
+    render() {
+        return (
+            <div className="all-over">
+                <Modal show={!this.state.user && this.state.showModal} onExited = {() => this.createChat()}>
```
Comment (nushicoco) [inline client/src/components/chat/chat.js]: yeah that's a major issue, that's why I was having so much trouble with getting the user aftr the modal has been mounted, a thing I can assume can be solved with redux.

File: client/src/components/chat/chat.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                                } } >{Strings.chat.startChat}
+                        </Button>
+                    </Modal.Body>
+
+                    <Modal.Footer>
+                        <BackToFrequentBtn/>
+                    </Modal.Footer>
+                </Modal>
+
+                { this.state.chats.length == 0 && this.state.user && this.state.user.isSuperuser &&
```
Comment (nushicoco) [inline client/src/components/chat/chat.js]: 👍

File: client/src/components/chat/chat.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                    </Tabs>
+                </div>
+                }
+
+                <div>
+                    <BackToFrequentBtn/>
+
+                    {/*{(!this.state.user || !this.state.user.isSuperuser) &&*/}
+                    {/*<NewTicketButton />*/}
+                    {/*}*/}
```
Comment (nushicoco) [inline client/src/components/chat/chat.js]: it's to remind us to put the button there someday but I can remove it

File: client/src/components/chat/chatTab.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        let self = this;
+
+        getChat(this.state.chatId).then(function (chat) {
+            self.setState({messages:chat.chat_messages});
+        })
+    }
+
+    componentWillMount() {
+        let self = this;
+        this.socket.emit(`client:connected`, { chatId : this.state.chatId , client: this.state.client, clientId:this.state.user && this.state.userId });
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: 👍

File: client/src/components/chat/chatTab.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                    {Strings.chat.chatWithTech}
+                </h2>
+            }
+            <div className="chat-area" >
+                {this.state.messages.map((msg, index) => {
+                    return <Message key={index}
+                                    author={msg.client}
+                                    text={msg.text}
+                                    time={msg.createdAt}
+                                    isMe={ msg.clientId == this.state.clientId  || msg.client == this.state.client}></Message>
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: b/c of an edge case when superuser use direct link and the client id gets to be the socket other than the user.id, and if he refresh again he's msgs will appear from someone else (different socket.id)

File: client/src/components/chat/chatTab.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+                })}
+                <div style={{ float:"left", clear: "both" }}
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: sure

File: client/src/components/chat/message.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            <div className={this.state.isMe ?  "arrow-right arrow" : "arrow-left arrow" }></div>
+            <div className={this.state.isMe ?  "right msg other-message" : "left msg my-message"}>
+                <span className="author">{this.state.author }</span>
+                <br/>
+                <span className="msg-txt">{this.state.text}</span>
+                <br/>
+                <span className="msg-time">{Moment(this.state.time).format("D/M/YY H:mm")}</span>
+            </div>
+            <div className="dvider">
+                <span>SomeLongTextHereSoWeWouldBreakALineMightBeAHackButIDontCare</span>
```
Comment (nushicoco) [inline client/src/components/chat/message.js]: lol :-) a lot of css teeth broke here

File: client/src/components/chat/textSubmitter.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+import {Row, Col, Form, FormGroup, FormControl, ControlLabel, Button} from 'react-bootstrap'
+
+import Strings from '../../strings';
+
+export default class TextSubmitter extends Component {
+    constructor (props) {
+        super(props);
+
+        this.state = {
+            chatId: props.chatId
```
Comment (nushicoco) [inline client/src/components/chat/textSubmitter.js]: just props?

File: client/src/components/common.js [hunk 15]
```diff
@@ -20,3 +27,17 @@ export class NewTicketButton extends React.Component {
...
     }
 }
+
+export class NewChatButton extends React.Component {
+
+    render () {
+        return (
+            <Link to="/chat" >
+                <Button className="open-chat">
+                    {/*<Glyphicon glyph="log-in" />*/}
```
Comment (nushicoco) [inline client/src/components/common.js]: weird!

File: client/src/index.css [hunk 16]
```diff
@@ -46,3 +42,34 @@ a:hover {
 .inline {
   display: inline;
 }
+
+.all-over {
+  width: 100%;
+  height: 100%;
+  margin: 0em;
```
Comment (nushicoco) [inline client/src/index.css]: It was copied from stackoverflow

File: io.js [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+            for (room in socket.rooms){
+                if (room == socket.id){
+                    continue;
+                }
+
+                // if no more clients in the room we close the chat
+                var clients = io.sockets.adapter.rooms[room];
+                if (clients && clients.length == 1){
+                    Chat.update({status:"closed"}, {where:{id:room}}).then(function (count){
+                        return;
```
Comment (nushicoco) [inline io.js]: used for console.log when debugging. I can remove it.

File: migrations/20170624135206-create-chat-and-chat-messages.js [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+              allowNull: false,
+              type: Sequelize.DATE
+          },
+          updatedAt: {
+              allowNull: false,
+              type: Sequelize.DATE
+          }
+      });
+
+      queryInterface.sequelize.query([
```
Comment (nushicoco) [inline migrations/20170624135206-create-chat-and-chat-messages.js]: couldn't find easily how to do it through the abstraction layer, do u know how?

File: migrations/20170624135206-create-chat-and-chat-messages.js [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+                                      FOREIGN KEY (chat_id) REFERENCES chats(id)
+                                      );`,
+              ].join(''),
+              { raw: true });
+  },
+
+  down: function (queryInterface, Sequelize) {
+
+      return queryInterface.dropTable('chats');
+      return queryInterface.dropTable('chatMessages');
```
Comment (nushicoco) [inline migrations/20170624135206-create-chat-and-chat-messages.js]: ?

File: models/chat.js [hunk 20]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Created by einavcarmon on 20/06/2017.
+ */
+const { notifyNewChat } = require('../notifications.js')
+const Sequelize = require('sequelize')
+
+module.exports = function (sequelize) {
+    const Chat = sequelize.define('chat', {
+        status: {
```
Comment (nushicoco) [inline models/chat.js]: yeah! much better.

File: models/chatMessage.js [hunk 21]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Created by einavcarmon on 20/06/2017.
+ */
+const Sequelize = require('sequelize')
+module.exports = function (sequelize) {
+    const ChatMessage = sequelize.define('chat_message', {
+        client: Sequelize.TEXT,
+        clientId: Sequelize.TEXT,
+        text: Sequelize.TEXT,
```
Comment (nushicoco) [inline models/chatMessage.js]: sure

File: routes/chat.js [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+module.exports = function (app, passport) {
+
+    const { Chat, ChatMessage, User  } = require('../models')
+
+    var counter = 0;
+    app.post('/chat', (req, res) => {
+        // todo create chat in db and query them
+        if (req.user && req.user.isSuperuser){
```
Comment (nushicoco) [inline routes/chat.js]: ok sounds good

File: routes/chat.js [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+            console.log("super user logged in");
```
Comment (nushicoco) [inline routes/chat.js]: console.log? I'll remove it

File: routes/chat.js [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+            Chat.findAll({where:{status:'active'}})
+                .then( all => res.send(all.map( item => item.toJSON())))
+        }else{
+
+            var client = req.param("client");
```
Comment (nushicoco) [inline routes/chat.js]: for some reason I thought node doesn't support that. I'll fix it.

File: routes/chat.js [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+            var clientId = req.param("clientId");
+            Chat.create({status:'active', client: client, clientId:clientId})
+                .then( (chat) => {
+                    res.status(200).send(chat);
+                    return null;
```
Comment (nushicoco) [inline routes/chat.js]: otherwise we get a warning in the log that the promise didn't finish

File: routes/chat.js [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+        Chat.findOne({
+            where: {
+                id: chatId
+            },
+            include: [{all: true, nested: true}],
+            order: [[ChatMessage, 'createdAt', 'ASC']]
+        })
+        .then( (chat) => {
+            if (!chat){
+                res.sendStatus(400).send({errorMessage: "couldn't find chat"})
```
Comment (nushicoco) [inline routes/chat.js]: sure, I'll check

File: server.js [hunk 27]
```diff
@@ -37,6 +37,13 @@ const passport = require("./passportAuthentication");
 app.use(passport.initialize());
 app.use(passport.session());
 
+// socket.io
+var server = require('http').Server(app);
```
Comment (nushicoco) [inline server.js]: 👍

File: client/src/components/chat/chat.js [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+
+        this.state = {
+            chats : [],
+            client: props.user ? props.user.firstName : '',
+            activeChatIndex: 0,
+            showModal:true,
+            user: props.user
+        };
+
+        // todo move url to .env (@dancar help ?)
```
Comment (nushicoco) [inline client/src/components/chat/chat.js]: when I try to use the process.env my url is http://undefined for some reason.

File: client/src/components/chat/chat.js [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+            this.setState({client:this.state.user.firstName});
+        }
+    }
+
+    closeModal() {
+        this.setState({ showModal: false });
+    }
+
+    createChat(){
+        var self = this;
```
Comment (nushicoco) [inline client/src/components/chat/chat.js]: .bind(this) ?

File: client/src/components/chat/chat.js [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+                                             socket={ this.socket } isSuperuser={this.state.user && this.state.user.isSuperuser}/>
+                                </Tab>
+                            )
+                        })}
+                    </Tabs>
+                </div>
+                }
+
+                <div>
+                    <BackToFrequentBtn/>
```
Comment (nushicoco) [inline client/src/components/chat/chat.js]: yeah, thought we can deal with that later.

File: client/src/components/chat/chatTab.js (hunk 10 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: that's a good point.
I did the testing on the serverside, I wonder if we could have this abstraction layer and use it on both ends?

File: client/src/components/chat/chatTab.js [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+        this.socket.on(`server:gotMessage`, data => {
+            // this is only for super user that should have one socket for all chats
+            if (data.chatId != this.state.chatId){
+                return;
+            }
+            self.setState( { messages : self.state.messages.concat([data]) } );
+        })
+    }
+
+    scrollToBottom = () => {
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: stackoverflow 😊

File: client/src/components/chat/chat.js [hunk 32]
```diff
@@ -0,0 +1,130 @@
...
+                    <Modal.Footer>
+                        <BackToFrequentBtn/>
+                    </Modal.Footer>
+                </Modal>
+
+                { this.state.chats.length == 0 && this.state.user && this.state.user.isSuperuser &&
+                    <p>No chats pending</p>
+                }
+                { this.state.chats.length == 1 &&
+                    <ChatTab chatId={this.state.chats[0].id} client={ this.state.client || this.state.user.firstName} socket={ this.socket }></ChatTab>
```
Comment (nushicoco) [inline client/src/components/chat/chat.js]: 👍

File: client/src/components/chat/chatTab.js [hunk 33]
```diff
@@ -0,0 +1,100 @@
...
+    componentWillMount() {
+        let self = this;
+        this.socket.emit(`client:connected`, { chatId : this.state.chatId , client: this.state.client, clientId:this.state.user && this.state.userId });
+        this.socket.on (`server:connected`, data => {
+            this.setState({clientId:data.clientId});
+        });
+
+        this.socket.on(`server:gotMessage`, data => {
+            // this is only for super user that should have one socket for all chats
+            if (data.chatId != this.state.chatId){
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: it turns out that my IDE has warnings I just ignored them :)

File: client/src/components/chat/chatTab.js [hunk 34]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    sendMessage = message => {
+        this.socket.emit(`client:sendMessage`, { text: message, chatId:this.state.chatId})
+    }
+
+    render() {
+        return (
+        <div className="mini-container">
+            {this.state.isSuperuser &&
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: 👍

File: client/src/components/chat/message.js [hunk 35]
```diff
@@ -0,0 +1,36 @@
...
+ * Created by einavcarmon on 19/06/2017.
+ */
+import React, { Component } from 'react';
+import Moment from 'moment';
+
+export default class Message extends Component {
+    constructor (props) {
+        super(props);
+
+        this.state = {
```
Comment (nushicoco) [inline client/src/components/chat/message.js]: 👍

File: client/src/components/chat/chatTab.js [hunk 36]
```diff
@@ -0,0 +1,100 @@
...
+                    return <Message key={index}
+                                    author={msg.client}
+                                    text={msg.text}
+                                    time={msg.createdAt}
+                                    isMe={ msg.clientId == this.state.clientId  || msg.client == this.state.client}></Message>
+                })}
+                <div style={{ float:"left", clear: "both" }}
+                     ref={(el) => { this.messagesEnd = el; }} />
+            </div>
+            <TextSubmitter sendMessage={ this.sendMessage.bind(this) }>
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: 👍

File: models/chatMessage.js (hunk 21 above)
Comment (nushicoco) [inline models/chatMessage.js]: did 15 for nickname and 200 for msg

File: server.js [hunk 37]
```diff
@@ -37,6 +37,13 @@ const passport = require("./passportAuthentication");
...
+server.listen(process.env.CHAT_PORT,process.env.CHAT_HOST);
```
Comment (nushicoco) [inline server.js]: ?

File: client/src/components/chat/chat.js (hunk 28 above)
Comment (dancar) [inline client/src/components/chat/chat.js]: Did you have a look at how we implemented env-vars in analytics.js? no reason it shouldn't work the same I reckon

File: client/src/components/chat/chat.js (hunk 29 above)
Comment (dancar) [inline client/src/components/chat/chat.js]: you're correctly using the fat arrow `=>` in the anonymous function below, hence no need to use `bind` or `self` (this is what the fat arrow do, other than looking awesome)

File: client/src/components/chat/chat.js (hunk 7 above)
Comment (dancar) [inline client/src/components/chat/chat.js]: I believe we can solve it without it, too

File: client/src/components/chat/chat.js (hunk 30 above)
Comment (dancar) [inline client/src/components/chat/chat.js]: 👍

File: client/src/components/chat/chatTab.js (hunk 10 above)
Comment (dancar) [inline client/src/components/chat/chatTab.js]: yea probably. We might have to hack the module system (I'm thinking symbolic links again) but it should work

File: client/src/components/chat/chatTab.js (hunk 11 above)
Comment (dancar) [inline client/src/components/chat/chatTab.js]: mmm I don't understand the system as well as you do, I only suspect there should be a more elegant solution. Do we really need to allow superusers to create chats themselves? and if so, why should it differ from how regular users do it?
I might be way off tho

File: client/src/components/chat/textSubmitter.js (hunk 14 above)
Comment (dancar) [inline client/src/components/chat/textSubmitter.js]: yup. might even wanna ditch the entire constructor

File: client/src/components/common.js (hunk 15 above)
Comment (dancar) [inline client/src/components/common.js]: I have it working on my other project. It might be something with the theme

File: client/src/index.css (hunk 16 above)
Comment (dancar) [inline client/src/index.css]: turns out it's a (the) font-relative size but can be used for other properties.

File: migrations/20170624135206-create-chat-and-chat-messages.js (hunk 18 above)
Comment (dancar) [inline migrations/20170624135206-create-chat-and-chat-messages.js]: How does it differ from how you made the 'chats' table?

File: migrations/20170624135206-create-chat-and-chat-messages.js (hunk 19 above)
Comment (dancar) [inline migrations/20170624135206-create-chat-and-chat-messages.js]: you're giving too much in return

File: models/chatMessage.js (hunk 21 above)
Comment (dancar) [inline models/chatMessage.js]: 👍

File: routes/chat.js (hunk 24 above)
Comment (nushicoco) [inline routes/chat.js]: it didn't work b/c param is a method.

File: routes/chat.js (hunk 23 above)
Comment (dancar) [inline routes/chat.js]: 👍  the tests thank you in advanced

File: server.js (hunk 37 above)
Comment (dancar) [inline server.js]: The host and port for the chat are a piece of information used both the server/ here, and by the client to reach it. Shouldn't we define them in one place that would be used by both sides?

File: client/public/index.html [hunk 38]
```diff
@@ -30,6 +30,17 @@
     <link rel="stylesheet" href="%PUBLIC_URL%/bootstrap-rtl.min.css">
     <!-- TODO: local copy?  -->
 
+    <!-- jaco - seeing user's use the system -->
+    <!--<script type="text/javascript">-->
+        <!--(function(e,r){function n(e,r){e[r]=function(){e.push([r].concat(Array.prototype.slice.call(arguments,0)))}}function o(){var e=r.location.hostname.match(/[a-z0-9][a-z0-9\-]+\.[a-z\.]{2,6}$/i),n=e?e[0]:null,o="; domain=."+n+"; path=/; expires=" + new Date(new Date().setFullYear(new Date().getFullYear() + 1)).toUTCString();r.cookie=r.referrer&&-1===r.referrer.indexOf(n)?"jaco_referer="+r.referrer+o:"jaco_referer="+t+o}var a="JacoRecorder",t="none";!function(e,r,t,i){if(!t.__VERSION){e[a]=t;for(var c=["init","identify","startRecording","stopRecording","removeUserTracking","setUserInfo","trackEvent"],s=0;s<c.length;s++)n(t,c[s]);o(),t.__VERSION=2.1,t.__INIT_TIME=1*new Date;var f=r.createElement("script");f.async=!0,f.setAttribute("crossorigin","anonymous"),f.src=i;var d=r.getElementsByTagName("head")[0];d.appendChild(f)}}(e,r,e[a]||[],"https://recorder-assets.getjaco.com/recorder_v2.js")}).call(window,window,document);-->
```
Comment (nushicoco) [inline client/public/index.html]: not sure, this is how they instruct the installation

File: routes/chat.js [hunk 39]
```diff
@@ -1,33 +1,35 @@
 /**
  * Created by einavcarmon on 20/06/2017.
  */
+const { onlySuperuser } = require('./helpers')
```
Comment (dancar) [inline routes/chat.js]: hmm maybe put all 'require's in the same place? not sure how that mechanism works so can't really say which one's faster (but pretty sure there's no functional difference since a module has its own scope)

File: client/src/components/chat/chat.js (hunk 7 above)
Comment (nushicoco) [inline client/src/components/chat/chat.js]: cool lets talk about it

File: client/src/components/chat/chatTab.js [hunk 40]
```diff
@@ -0,0 +1,100 @@
...
+
+import Strings from '../../strings';
+
+
+export default class chatTab extends Component {
+
+    constructor (props) {
+        super(props);
+
+        this.state = {
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: 👍

File: client/src/components/chat/chatTab.js [hunk 41]
```diff
@@ -0,0 +1,100 @@
...
+            chatId : props.chatId,
+            messages: [],
+            client: props.client,
+            clientId: '',
+            isSuperuser: props.isSuperuser
+        }
+
+        this.socket = props.socket;
+
+        let self = this;
```
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: will be fixed in the next version

File: client/src/components/chat/textSubmitter.js (hunk 14 above)
Comment (nushicoco) [inline client/src/components/chat/textSubmitter.js]: next branch

File: migrations/20170624135206-create-chat-and-chat-messages.js (hunk 18 above)
Comment (nushicoco) [inline migrations/20170624135206-create-chat-and-chat-messages.js]: the foreign key
